trieste/utils/pareto.py

```diff
@@ -0,0 +1,309 @@
+            tf.logical_and(
+                tf.reduce_all(extended <= swapped_ext, axis=2),
+                tf.reduce_any(extended < swapped_ext, axis=2),
+            ),
+            tf.float32,
+        ),
+        axis=1,
+    )
+
+    return tf.boolean_mask(observations, dominance == 0), dominance
+
+
+class BoundedVolumes:
+    def __init__(self, lb: tf.Tensor, ub: tf.Tensor):
+        """
+        Construct bounded volumes.
+        :param lb: the lowerbounds of the volumes
+        :param ub: the upperbounds of the volumes
+        """
+
+        assert tf.reduce_all(lb.shape == ub.shape)
+        self.lb = lb
+        self.ub = ub
+
+    def append(self, lb: tf.Tensor, ub: tf.Tensor):
+        """
+        Add new bounded volumes.
+        :param lb: the lowerbounds of the volumes
+        :param ub: the upperbounds of the volumes
+        """
+        self.lb = tf.concat([self.lb, lb], axis=0)
+        self.ub = tf.concat([self.ub, ub], axis=0)
+
+    def clear(self):
+        """
+        Clears all stored bounded volumes
+        """
+        dtype = self.lb.dtype
+        outdim = self.lb.shape[1]
+        self.lb = tf.zeros([0, outdim], dtype=dtype)
+        self.ub = tf.zeros([0, outdim], dtype=dtype)
+
+    def size(self) -> tf.Tensor:
+        """
+        :return: volume of each bounded volume
+        """
+        return tf.math.reduce_prod(self.ub - self.lb, axis=0, keepdims=True)
+
+
+class Pareto:
+    def __init__(self, datasets: Mapping[str, Dataset], threshold: tf.Tensor = 0):
+        """
+        Construct a Pareto set.
+        Stores a Pareto set and calculates the cell bounds covering the non-dominated region.
+        The latter is needed for certain multiobjective acquisition functions.
+        :param datasets: Mapping of Dataset and its coresponding string key(s)
+        :param threshold: approximation threshold for the generic divide and conquer strategy
+            (default 0: exact calculation)
+        """
+        self.threshold = threshold
+        self.data = to_Dataset(datasets)
+
+        # Setup data structures
+        self.bounds = BoundedVolumes(
+            tf.zeros([0, self.data.observations.shape[1]], dtype=tf.int32),
+            tf.zeros([0, self.data.observations.shape[1]], dtype=tf.int32),
+        )
+
+        self.front = tf.zeros([0, self.data.observations.shape[1]])
+        # Initialize
+        self.update()
+
+    @staticmethod
+    def _is_test_required(smaller: tf.Tensor) -> tf.Tensor:
+        """
+        Tests if a point augments or dominates the Pareto set.
+        :param smaller: a boolean tf.Tensor storing test point < Pareto front
+        :return: True if the test point dominates or augments the Pareto front (boolean)
+        """
+        # if and only if the test point is at least in one dimension smaller
+        # for every point in the Pareto set
+        idx_dom_augm = tf.reduce_any(smaller, axis=1)
+        is_dom_augm = tf.reduce_all(idx_dom_augm)
```

Comment:
there are a lot of very short names. It would help readability if they were more descriptive